text/0037-node-upgrades.md

```diff
@@ -0,0 +1,98 @@
+they become active.
+
+The node has two relevant modes of operation: "validator mode", and "migrate-data mode".  The normal mode is validator.
+The migrate-data mode is used immediately after an upgrade to migrate any data stored on disk as required.
+
+The node also has three possible exit codes: 0, 101 or 102.  0 indicates success, 101 indicates an error (e.g. a panic)
+and 102 indicates the node should be downgraded to a previous version.  These exit codes are read and reacted to by the
+launcher.
+
+On startup, the launcher either tries to read its previously cached state from disk, or assumes a fresh start.
+Information cached includes which mode of operation the node was executed in, the protocol version, the path to the node
+binary and the path to the config folder.
+
+On a fresh start, the launcher searches for the lowest installed version of `casper-node` and starts running it in
+validator mode.
+
+If the launcher cannot find an appropriate version at any stage of upgrading or downgrading, it exits with an error.
+
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+### casper-node-launcher
+
+The following describes the control flow in the launcher:
+
+![control flow in the launcher](images/0037/casper-launcher-flow.png "control flow in the launcher")
+
+### casper-node
+
+On startup, the node needs to decide what action to take depending upon whether it is an initial run (without fast sync
+this will always be treated as running from genesis), a restart after an upgrade, or a restart after an unexpected
+shutdown.  The previous run (if any) is deduced by examining the stored linear chain, specifically by querying the
+highest block, and comparing that to the current and next (if available) activation points.
+
+Note that if a node starts with a chainspec which specifies `protocol.hard_reset = true`, then all blocks created in or
+after the chainspec's activation point are effectively removed from storage, creating a situation as if the node had
+ceased operating at the end of the appropriate era.  This option is seen as a failsafe, to cater for the scenario where
+the network has ceased operating properly mid-era.
+
+The actions a node will take on startup, in the order in which they are tested, are summarized below:
+
+* HB = highest block
+* PV = protocol version
+* AP = activation point
+* current X = value taken from current chainspec
+* next X = value taken from next upgraded chainspec
+
+| Condition                                           | Scenario                                     | Action                                   |
+| :-------------------------------------------------- | :------------------------------------------- | :--------------------------------------- |
+| no blocks in storage, and PV 1.0.0                  | initial run                                  | commit genesis and continue running      |
```

Comment:
In the future, we could add an optional "starting protocol version" to the chainspec, so you could e.g. launch a private network without having to start from 1.0.0.

Reply:
Well, I can't really see a good reason for wanting that, but it's certainly do-able :)  Either way, it's probably outside the scope of this CEP, since it's not really related to network upgrades?